### PR TITLE
Feature/vec 2 x

### PR DESCRIPTION
**Pull request by TobiasFehlmann**
## Pull Request

- [x] I have checked for similar PRs.
- [x] I have read the [contributing guidelines](https://github.com/4Soft-de/harness-model/blob/develop/.github/CONTRIBUTING.md).

### Changes

- [x] Code
- [x] Documentation
- [ ] Other: 

Closes: \<Put down the issue number if available or remove that line>

### Description

Adds a new module for converting between VEC 1.2.x and 2.0.x

### CODE REVIEW

File: compatibility/compatibility-vec12to20/src/main/java/com/foursoft/harness/compatibility/vec12to20/util/InitializeFields.java [hunk 1]
```diff
@@ -0,0 +1,182 @@
+/*-
+ * ========================LICENSE_START=================================
+ * Compatibility VEC 1.1.X To VEC 1.2.X
```
Comment (DManstrator) [inline compatibility/compatibility-vec12to20/src/main/java/com/foursoft/harness/compatibility/vec12to20/util/InitializeFields.java:3]: Header was not adjusted.

File: compatibility/compatibility-vec12to20/src/main/java/com/foursoft/harness/compatibility/vec12to20/util/Vec20XReader.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+import com.foursoft.harness.compatibility.vec.common.VecCreator;
+import com.foursoft.harness.compatibility.vec.common.VecProcessor;
+import com.foursoft.harness.compatibility.vec12to20.Vec12XTo20XProcessor;
+
+import java.io.InputStream;
+
+/**
+ * Class which is able to read and create any VEC 1.X.X.
+ */
```
Comment (DManstrator) [inline compatibility/compatibility-vec12to20/src/main/java/com/foursoft/harness/compatibility/vec12to20/util/Vec20XReader.java:36]: JavaDocs were not adjusted.

File: compatibility/compatibility-vec12to20/src/main/java/com/foursoft/harness/compatibility/vec12to20/util/Vec20XReader.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+     *
+     * @param inputStream InputStream of a VEC.
+     * @param streamName  The name of the stream for logging.
+     * @return The VecContent in version 1.2.X.
+     */
+    public static com.foursoft.harness.vec.v12x.
+            VecContent createVec12x(final InputStream inputStream, final String streamName) {
+        return new VecCreator(CREATOR).createVec(inputStream, streamName,
+                                                 com.foursoft.harness.vec.v12x.VecContent.class);
+    }
```
Comment (DManstrator) [inline compatibility/compatibility-vec12to20/src/main/java/com/foursoft/harness/compatibility/vec12to20/util/Vec20XReader.java:69]: In my opinion, this should come first / the methods should be in the same order (vec12, vec2, vec12, vec2).

File: compatibility/compatibility-vec12to20/src/main/java/com/foursoft/harness/compatibility/vec12to20/wrapper/vec12to20/Vec12To20CableLeadThroughWrapper.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
+/*-
+ * ========================LICENSE_START=================================
+ * Compatibility VEC 1.1.X To VEC 1.2.X
```
Comment (DManstrator) [inline compatibility/compatibility-vec12to20/src/main/java/com/foursoft/harness/compatibility/vec12to20/wrapper/vec12to20/Vec12To20CableLeadThroughWrapper.java:3]: Header was not adjusted (probably in all Wrappers).

File: compatibility/compatibility-vec12to20/src/main/java/com/foursoft/harness/compatibility/vec12to20/wrapper/vec12to20/Vec12To20DocumentVersionWrapper.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+     * Creates this wrapper.
+     *
+     * @param context Context of the wrapper.
+     * @param target  Target object of the wrapper.
+     */
+    public Vec12To20DocumentVersionWrapper(final CompatibilityContext context, final Object target) {
+        super(context, target);
+    }
+
+    protected BigInteger numberOfSheets;
```
Comment (DManstrator) [inline compatibility/compatibility-vec12to20/src/main/java/com/foursoft/harness/compatibility/vec12to20/wrapper/vec12to20/Vec12To20DocumentVersionWrapper.java:55]: The field(s) should come before the constructor (probably in more Wrappers too).

File: compatibility/compatibility-vec12to20/src/main/java/com/foursoft/harness/compatibility/vec12to20/wrapper/vec20to12/Vec20To12PartOccurrenceWrapper.java [hunk 6]
```diff
@@ -23,37 +23,26 @@
  * THE SOFTWARE.
  * =========================LICENSE_END==================================
  */
-package com.foursoft.harness.compatibility.vec11to12.util;
+package com.foursoft.harness.compatibility.vec12to20.wrapper.vec20to12;
```
Comment (DManstrator) [inline compatibility/compatibility-vec12to20/src/main/java/com/foursoft/harness/compatibility/vec12to20/wrapper/vec20to12/Vec20To12PartOccurrenceWrapper.java:26]: History is correct, GitHub is messing that up for some reason.

File: compatibility/compatibility-vec12to20/src/test/resources/logback-test.xml [hunk 7]
```diff
@@ -0,0 +1,20 @@
+<configuration debug="false">
+    <appender name="STDOUT" class="ch.qos.logback.core.ConsoleAppender">
+        <encoder class="ch.qos.logback.classic.encoder.PatternLayoutEncoder">
+            <Pattern>%d{HH:mm:ss.SSS} - %-10.10thread - %-5level - %-25.25logger{25} - %msg -- at .\(%F:%L\)%n</Pattern>
+        </encoder>
+    </appender>
+
+    <logger name="org.springframework" level="warn"/>
```
Comment (DManstrator) [inline compatibility/compatibility-vec12to20/src/test/resources/logback-test.xml:8]: Why is that here again?

File: vec/vec-v113/src/main/java/com/foursoft/harness/vec/v113/predicates/VecPredicates.java [hunk 8]
```diff
@@ -0,0 +1,250 @@
+/*-
+ * ========================LICENSE_START=================================
+ * VEC 2.X
```
Comment (DManstrator) [inline vec/vec-v113/src/main/java/com/foursoft/harness/vec/v113/predicates/VecPredicates.java:3]: Wrong header.

File: vec/vec-v113/src/main/java/com/foursoft/harness/vec/v113/predicates/VecPredicates.java [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+import com.foursoft.harness.vec.v113.*;
+
+import java.math.BigInteger;
+import java.util.Objects;
+import java.util.function.Predicate;
+
+/**
+ * Uncategorized {@link Predicate} methods.
+ */
+public final class VecPredicates {
```
Comment (DManstrator) [inline vec/vec-v113/src/main/java/com/foursoft/harness/vec/v113/predicates/VecPredicates.java:38]: Honestly I'm not sure if we really need that class. I can see that it should be consistent but I'm still not sure if we should have them when there is already a VEC 2.X.

File: vec/vec-v2x/pom.xml [hunk 10]
```diff
@@ -27,6 +27,11 @@
             <groupId>jakarta.annotation</groupId>
             <artifactId>jakarta.annotation-api</artifactId>
         </dependency>
+        <dependency>
+            <groupId>com.foursoft.harness.vec</groupId>
+            <artifactId>vec-v12x</artifactId>
+            <version>2.2.0</version>
+        </dependency>
```
Comment (DManstrator) [inline vec/vec-v2x/pom.xml:34]: Why is that needed? And why in that version?

**Comment by RainerGanss** [comment]
@tobyfehlmann What about this PR? Would you please comment on the feedback?

File: vec/vec-v2x/pom.xml (hunk 10 above)
Comment (TobiasFehlmann) [reply vec/vec-v2x/pom.xml]: removed

File: vec/vec-v12x/src/main/java/com/foursoft/harness/vec/v12x/predicates/VecPredicates.java [hunk 11]
```diff
@@ -53,18 +55,43 @@ public static Predicate<VecPlaceableElementRole> onPointPlacement() {
                 .anyMatch(VecPlacementType.ON_POINT::equals);
     }
 
+    /**
+     * Checks if the given localized string is german.
+     *
+     * @return check result.
+     */
```
[the remark this answers is not in the document]
Comment (TobiasFehlmann) [reply vec/vec-v12x/src/main/java/com/foursoft/harness/vec/v12x/predicates/VecPredicates.java]: I removed the @return as like in the other ones.

File: vec/vec-v113/src/main/java/com/foursoft/harness/vec/v113/predicates/VecPredicates.java (hunk 8 above)
Comment (TobiasFehlmann) [reply vec/vec-v113/src/main/java/com/foursoft/harness/vec/v113/predicates/VecPredicates.java]: updated

File: compatibility/compatibility-vec12to20/src/test/resources/logback-test.xml (hunk 7 above)
Comment (TobiasFehlmann) [reply compatibility/compatibility-vec12to20/src/test/resources/logback-test.xml]: removed

File: compatibility/compatibility-vec11to12/pom.xml [hunk 12]
```diff
@@ -12,6 +12,11 @@
     <description>The VEC 1.1.* to 1.2.* layer.</description>
 
     <dependencies>
+        <dependency>
+            <groupId>${project.groupId}</groupId>
+            <artifactId>compatibility-vec-common</artifactId>
+            <version>${project.version}</version>
+        </dependency>
```
[the remark this answers is not in the document]
Comment (TobiasFehlmann) [reply compatibility/compatibility-vec11to12/pom.xml]: removed

File: compatibility/compatibility-vec-common/src/test/java/com/foursoft/harness/compatibility/vec/common/util/VecVersionDetectorTest.java [hunk 13]
```diff
@@ -25,6 +25,8 @@
  */
 package com.foursoft.harness.compatibility.vec.common.util;
 
+import com.foursoft.harness.compatibility.vec.common.VecVersion;
+import com.foursoft.harness.compatibility.vec.common.util.VecVersionDetector;
```
[the remark this answers is not in the document]
Comment (TobiasFehlmann) [reply compatibility/compatibility-vec-common/src/test/java/com/foursoft/harness/compatibility/vec/common/util/VecVersionDetectorTest.java]: removed

File: compatibility/compatibility-vec-common/src/main/java/com/foursoft/harness/compatibility/vec/common/VecVersion.java [hunk 14]
```diff
@@ -48,7 +48,12 @@ public enum VecVersion {
     /**
      * Enum value for the VEC 1.2.X.
      */
-    VEC12X("1.2.X", "1.2.0");
+    VEC12X("1.2.X", "1.2.0"),
+
+    /**
+     * Enum value for the VEC 2.0.X.
+     */
+    VEC20X("2.0.X", "2.0.2");
```
[the remark this answers is not in the document]
Comment (TobiasFehlmann) [reply compatibility/compatibility-vec-common/src/main/java/com/foursoft/harness/compatibility/vec/common/VecVersion.java]: changed everywhere

File: compatibility/compatibility-vec-common/src/main/java/com/foursoft/harness/compatibility/vec/common/VecCreator.java [hunk 15]
```diff
@@ -48,7 +48,8 @@ public final class VecCreator {
 
     private static final Map<String, VecVersion> CLASS_PATH_TO_VEC_VERSION_MAP = Map.ofEntries(
             Map.entry("com.foursoft.harness.vec.v113.VecContent", VecVersion.VEC11X),
-            Map.entry("com.foursoft.harness.vec.v12x.VecContent", VecVersion.VEC12X)
+            Map.entry("com.foursoft.harness.vec.v12x.VecContent", VecVersion.VEC12X),
+            Map.entry("com.foursoft.harness.vec.v2x.VecContent", VecVersion.VEC12X)
```
[the remark this answers is not in the document]
Comment (TobiasFehlmann) [reply compatibility/compatibility-vec-common/src/main/java/com/foursoft/harness/compatibility/vec/common/VecCreator.java]: fixed

File: compatibility/compatibility-vec12to20/src/main/java/com/foursoft/harness/compatibility/vec12to20/Vec12XTo20XClassMapper.java [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+     * Creates a VEC 1.1 to VEC 1.2 Class Mapper.
+     */
+    public Vec12XTo20XClassMapper() {
+        super(Constants.PACKAGE_VEC12X, Constants.PACKAGE_VEC20X);
+
+        classMap = new HashMap<>();
+
+        // VEC 1.2.X -> VEC 2.0.X
+
+        // VEC 2.0.X -> VEC 1.2.X
```
[the remark this answers is not in the document]
Comment (TobiasFehlmann) [reply compatibility/compatibility-vec12to20/src/main/java/com/foursoft/harness/compatibility/vec12to20/Vec12XTo20XClassMapper.java]: I don't get your point. The ClassMapper need dependencies to it's 2 VEC versions for conversion. But there is no module with dependencies to all 3 versions. And this refectoring is not enough to do so in my opinion.

[the remark this answers is not in the document]
Comment (DManstrator) [reply compatibility/compatibility-vec12to20/src/main/java/com/foursoft/harness/compatibility/vec12to20/Vec12XTo20XClassMapper.java]: Currently, `classMap` is empty since there is no addition to it. Due to that, the `map` method also always executes the else case. In general, it just returns `super.map(clazz)`. So I would either simplify it to that or refactor it so there is just one `VecClassMapper`. For example it could take the "conversion maps" as constructor parameters. Because so fair, `classMap` uses wildcards and `UnsupportedVec12XToVec20XMethods` has a `Map<MethodIdentifier>`.

And I would prefer the refactoring.

File: vec/vec-v12x/src/main/java/com/foursoft/harness/vec/v12x/predicates/VecPredicates.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (DManstrator) [reply vec/vec-v12x/src/main/java/com/foursoft/harness/vec/v12x/predicates/VecPredicates.java]: I still don't get why there are JavaDocs for just half the methods. Could you add them for the other methods too?

File: compatibility/compatibility-vec12to20/src/main/java/com/foursoft/harness/compatibility/vec12to20/Vec12XTo20XProcessor.java [hunk 17]
```diff
@@ -0,0 +1,142 @@
+/*-
+ * ========================LICENSE_START=================================
+ * Compatibility VEC 1.1.X To VEC 1.2.X
```
[the remark this answers is not in the document]
Comment (DManstrator) [reply compatibility/compatibility-vec12to20/src/main/java/com/foursoft/harness/compatibility/vec12to20/Vec12XTo20XProcessor.java:3]: There are still 17 unresolved comments, see the changed files. Can you look into them?

File: compatibility/compatibility-vec12to20/src/main/java/com/foursoft/harness/compatibility/vec12to20/Vec12XTo20XClassMapper.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (beckerjohannes) [reply compatibility/compatibility-vec12to20/src/main/java/com/foursoft/harness/compatibility/vec12to20/Vec12XTo20XClassMapper.java]: @DManstrator you are right, there is no addition to Map. If the intention of the `classMap` is to provide a "cache" of the calculated values, then are some problems with that:

1. @TobiasFehlmann Maybe this is a misunderstanding of the `Map.getOrDefault` method. This method returns a default value, if there is no value for the key in the map. However, the calculated value is not placed in the map. It is a bassically a "substitution" method for not returning null. E.g. in a `xy->String` map for translations the default might be `"Not Found"`.
2. If using `getOrDefault`, the default value should be a constant. Since the parameter is passed as a value the `super.map(clazz)` method is called always, regardless if its value is required or not. So, even if `getOrDefault` would be a valid caching approach, it would be destroyed be the call to `super.map(clazz)` because this is executed every time line 76 is executed.
3. When you want to use a map as cache you do the following:
```java 
   public Class<?> map(final Class<?> clazz) {
        final Class<?> aClass = classMap.get(clazz);
        return aClass != null ? aClass : classMap.computeIfAbsent(ClassUtils.getNonProxyClass(clazz), super::map);
    }
```

[the remark this answers is not in the document]
Comment (DManstrator) [reply compatibility/compatibility-vec12to20/src/main/java/com/foursoft/harness/compatibility/vec12to20/Vec12XTo20XClassMapper.java]: The `classMap` has the use to extend the mapping of classes from one package to another package.

In most cases, a class (e.g. `VecPartOccurrence`) is present in both packages. So the basic code is "replace the source package with the target package and try to find the class".

Example: `com.foursoft.harness.vec.v113.VecPartOccurrence` -> `com.foursoft.harness.vec.v12x.VecPartOccurrence`. Exists? Yes.

 However there are classes where this is not the case, e.g. VEC 1.1.X having `VecZone` and VEC 1.2.X having `VecTopologyZone` instead. If the default `map` method would be invoked, an exception would be thrown since the class can't be found in the other package.

I added the `getOrDefault` part when I recognized that `clazz` was sometimes a proxy class and sometimes not. So sometimes the class could be found in the `classMap`, sometimes not. So I added that as a fail safe.

So what is happening is basically this:

1. Is `clazz` present in `classMap`? (Checks if the given class needs a different class mapping.)
2. If yes, then simply return the mapped class provided by `classMap`.
3. If no, check: Is the non-proxy class of `clazz` is present in `classMap`? (Failsafe check.)
4. If yes, then simply return the mapped class provided by `classMap`.
5. If no, execute the super implementation.

Since this is confusing, I would add comments to the code or making it more verbose to make it more clear.

[the remark this answers is not in the document]
Comment (beckerjohannes) [reply compatibility/compatibility-vec12to20/src/main/java/com/foursoft/harness/compatibility/vec12to20/Vec12XTo20XClassMapper.java]: @DManstrator the problem is not that it is confusing what happens, that is quite clear. However, the point is, that it is sloppy programming style, because (as mentioned above) `super.map(clazz)` is executed always when line 76 is entered, no matter if the result is required or not. This is wasted execution time and can be, depending on the scenario quite expensive. In this this case, I don't think it really matters. I'll post a suggustion for a much clean solution.

File: compatibility/compatibility-vec12to20/src/main/java/com/foursoft/harness/compatibility/vec12to20/Vec12XTo20XClassMapper.java [hunk 18]
```diff
@@ -32,17 +32,14 @@
...
-import java.util.Objects;
+import java.util.*;
 
 /**
  * Class responsible for mapping VEC 1.2.X classes to VEC 2.X.X <b>and vice versa</b>.
  */
 public class Vec12XTo20XClassMapper extends NameBasedClassMapper {
 
-    private final Map<Class<?>, Class<?>> classMap;
+    private final Map<Class<?>, Class<?>> explicitClassMappings;
```
Comment (beckerjohannes) [inline compatibility/compatibility-vec12to20/src/main/java/com/foursoft/harness/compatibility/vec12to20/Vec12XTo20XClassMapper.java:42]: @DManstrator give at consice name that describes what it is for instead of generic one.

File: compatibility/compatibility-vec12to20/src/main/java/com/foursoft/harness/compatibility/vec12to20/Vec12XTo20XClassMapper.java [hunk 19]
```diff
@@ -70,10 +67,12 @@ public Vec12XTo20XClassMapper() {
...
-        final Class<?> aClass = classMap.get(clazz);
-        return aClass != null
-                ? aClass
-                : classMap.getOrDefault(ClassUtils.getNonProxyClass(clazz), super.map(clazz));
+        return resolveExplicitClassMappings(clazz).orElseGet(() -> super.map(clazz));
+    }
+
+    private Optional<Class<?>> resolveExplicitClassMappings(final Class<?> clazz) {
+        return Optional.<Class<?>>ofNullable(explicitClassMappings.get(clazz))
+                .or(() -> Optional.ofNullable(explicitClassMappings.get(ClassUtils.getNonProxyClass(clazz))));
```
Comment (beckerjohannes) [inline compatibility/compatibility-vec12to20/src/main/java/com/foursoft/harness/compatibility/vec12to20/Vec12XTo20XClassMapper.java:75]: @DManstrator extract functional units into small methods, give them consice names. Then you don't need inline comments.

**Comment by RainerGanss** [comment]
@tobyfehlmann What is the status of this PR? @takebene there is still a review requested from you.

**Comment by TobiasFehlmann** [comment]
> @tobyfehlmann What is the status of this PR? @takebene there is still a review requested from you.

I'm done. I added some tests to increase the coverage. But there are issues handeled in the wrappers where the case is not in the 'old beetle' test data.

**Review by RainerGanss — APPROVED** [review]

File: compatibility/compatibility-vec12to20/src/main/java/com/foursoft/harness/compatibility/vec12to20/wrapper/vec12to20/Vec12To20DocumentVersionWrapper.java [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+    protected Object wrapObject(final Object obj, final Method method, final Object[] allArguments) throws Throwable {
+        final String methodName = method.getName();
+        if ("getNumberOfSheets".equals(methodName)) {
+            if (numberOfSheets == null) {
+                handleValue(getResultObject("getNumberOfSheets", String.class).orElse("0"));
+            }
+            return numberOfSheets;
+        } else if ("setNumberOfSheets".equals(methodName)) {
+            numberOfSheets = (BigInteger) allArguments[0];
+            return null;
```
[the remark this answers is not in the document]
Comment (DManstrator) [reply compatibility/compatibility-vec12to20/src/main/java/com/foursoft/harness/compatibility/vec12to20/wrapper/vec12to20/Vec12To20DocumentVersionWrapper.java:67]: We need to check if what I mentioned is even needed or if it's enough like in this example. Will be done later.

File: compatibility/compatibility-vec12to20/src/main/java/com/foursoft/harness/compatibility/vec12to20/Vec12XTo20XClassMapper.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (DManstrator) [reply compatibility/compatibility-vec12to20/src/main/java/com/foursoft/harness/compatibility/vec12to20/Vec12XTo20XClassMapper.java]: The refactoring has still to be done, will happen in a later PR.

File: compatibility/compatibility-vec12to20/src/main/java/com/foursoft/harness/compatibility/vec12to20/util/InitializeFields.java [hunk 21]
```diff
@@ -0,0 +1,182 @@
...
+                        Proxy.newProxyInstance(
+                                InitializeFields.class.getClassLoader(),
+                                new Class[]{com.foursoft.harness.vec.v12x.visitor.Visitor.class},
+                                new DynamicInvocationHandler());
+
+        content.accept(
+                new com.foursoft.harness.vec.v12x.visitor.TraversingVisitor<>(
+                        new com.foursoft.harness.vec.v12x.visitor.DepthFirstTraverserImpl<>(),
+                        proxyInstance));
+    }
```
[the remark this answers is not in the document]
Comment (beckerjohannes) [reply compatibility/compatibility-vec12to20/src/main/java/com/foursoft/harness/compatibility/vec12to20/util/InitializeFields.java:77]: > Refactoring this isn't that easy. Either a module needs to know all VEC versions (which isn't really good) or the full class has to be passed as a String, which isn't really safe. We need to think about a proper solution later on.

Couldn't that be done with "Reflection only". Just an Idea...

1. You get in an Object. 
2. You check if it implements an Interface with simple name "Visitable".
3. If not -> ArgumentException
4. From the package to which `Visitable` belongs, you can derive the package of `Visitor`, `TraversingVisitor` and `DepthFirstTraverserImpl` and you can instantiate them an call `accept(...) purely based on reflection. 

No Classes or Strings required.
